### ISSUE TITLE
Issue with soundness of HLIL control flow structuring

### ISSUE THREAD

**Issue by plafosse** [Vector35/binaryninja-api#5201]
HLIL can produce unsound control flow structuring in some conditions.

Consider this MLIL code:

[orig_state_exec.zip](https://github.com/Vector35/binaryninja-api/files/14670193/orig_state_exec.zip)

In MLIL everything looks correct:
<img width="877" alt="image" src="https://github.com/Vector35/binaryninja-api/assets/10556585/fb0c677d-21dd-4b12-959f-70388115a9cc">

Consider the case when 'i = 0` the path means `i` is assigned to `5` and then ultimately goes to instruction `21`

Now in HLIL:
<img width="322" alt="image" src="https://github.com/Vector35/binaryninja-api/assets/10556585/8cd35134-69ac-43c3-a860-4b6aa5165ed7">
The control flow is a series of `if` statements rather than `if-else` statements. In the case of `i == 0` it meets the first condition and sets `i = 5` and then can also satisfy the second condition too incorrectly setting `var_20 = 1`

Special Thanks to: [Zao Yang](https://www.linkedin.com/in/zao-yang-8b7113171/) and [Stefan Nagy](https://users.cs.utah.edu/~snagy/) for their research in Decompiler Fuzzing for reporting this issue.

**Comment by D0ntPanic** [comment]
This is exposing a fundamental issue in control flow restructuring. We have desire to revisit the control flow restructuring algorithm to significantly reduce the number of conditions that get out of hand and become unreadable, which is also an issue with the current algorithm. This bug will probably have to be fixed alongside a refactor of the control flow restructuring itself.

**Comment by xusheng6** [comment]
A user reported a similar case:

- BN version: 4.0.4958 Windows
- Compiler: MSVC 2023 x64

Compile the following C code using `cl /c /O2 test.c` and disassemble the resulting binary in Binary Ninja:

```c
bool test(int *input)
{
    bool result;
    if(input[0] == 1) {
        result = true;
        input[0] = 2;
        if(input[1] == 1)
            goto other;
    } else {
other:
        result = false;
    }
    return result;
}
```

This results in the following High Level IL output:

```c
 int64_t test(int32_t* arg1) {
1    int64_t rax;
2    if (*arg1 == 1)
3        rax = 1;
4        *arg1 = 2;
5    if (*arg1 != 1 || (*arg1 == 1 && arg1[1] == 1))
6        rax.b = 0;
7    return rax;
 }
```

However, this output is not actually correct: if the function's input is `{ 1, 0 }`, at line 4 in the HLIL, `arg1[0]` will be set to 2. Then, at line 5, the condition in the `if` statement evaluates to `true` (since `arg1[0]` now evaluates to 2), which in turn leads to line 6 being executed, and the whole function returning `false`. This differs from the actual behavior of the program (which should be returning `true`).

In contrast, the Hex-Rays Decompiler generates the following output, which is correctly in correspondence to the actual semantics of the program:

```c
char __fastcall test(int *a1)
{
  char result; // al

  if ( *a1 != 1 || (result = 1, *a1 = 2, a1[1] == 1) )
    result = 0;
  return result;
}
```

I see the fundamental issue is the body of the if statement is modifying the variables used in the condition, so it alters the result

**Comment by D0ntPanic** [comment]
Fixed in 4.1.5547
